File: backend/app/schemas/account.py
from typing import Optional
from uuid import UUID
from datetime import date, datetime
from decimal import Decimal
from pydantic import BaseModel, ConfigDict, Field
from backend.app.models.account import AccountTypeEnum, AccountStatusEnum
from backend.app.schemas.institution import InstitutionRead


class AccountBase(BaseModel):
    institution_id: Optional[UUID] = None
    account_name: str
    account_type: AccountTypeEnum = AccountTypeEnum.CREDIT_CARD
    card_number_masked: str
    card_number_last4: str
    credit_limit: Decimal = Field(default=Decimal("0.00"))
    billing_day_of_month: Optional[int] = Field(default=None, ge=1, le=31)
    grace_period_days: int = Field(default=15, ge=0)
    status: AccountStatusEnum = AccountStatusEnum.ACTIVE
    replaces_account_id: Optional[UUID] = None
    opened_date: Optional[date] = None
    closed_date: Optional[date] = None
    color_hex: Optional[str] = "#3b82f6"
    note: Optional[str] = None


class AccountCreate(AccountBase):
    pass


class AccountUpdate(BaseModel):
    account_name: Optional[str] = None
    credit_limit: Optional[Decimal] = None
    billing_day_of_month: Optional[int] = None
    grace_period_days: Optional[int] = None
    status: Optional[AccountStatusEnum] = None
    replaces_account_id: Optional[UUID] = None
    closed_date: Optional[date] = None
    color_hex: Optional[str] = None
    note: Optional[str] = None


class AccountRead(AccountBase):
    id: UUID
    user_id: Optional[UUID] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    institution: Optional[InstitutionRead] = None

    model_config = ConfigDict(from_attributes=True)


class AccountOverviewRead(BaseModel):
    account_id: UUID
    account_name: str
    bank_name: str
    card_number_masked: str
    credit_limit: Decimal
    latest_statement_balance: Decimal
    next_payment_due_date: date
    status: AccountStatusEnum
    replaces_account_id: Optional[UUID] = None
    replaced_by_card_number: Optional[str] = None


class AccountLiveBalanceRead(BaseModel):
    account_id: UUID
    account_name: str
    bank_name: Optional[str] = None
    card_number_masked: str
    credit_limit: Decimal
    latest_statement_date: Optional[date] = None
    latest_statement_balance: Decimal
    unbilled_charges: Decimal
    unbilled_credits: Decimal
    unbilled_net_amount: Decimal
    unbilled_transaction_count: int
    live_current_balance: Decimal
    live_available_limit: Decimal
    live_utilization_percentage: Decimal
    live_risk_level: str
    next_payment_due_date: Optional[date] = None
    status: AccountStatusEnum

File: frontend/src/types/account.ts
import { Institution } from "./institution";

export type AccountType = "CREDIT_CARD" | "DEBIT_CARD" | "BANK_ACCOUNT" | "E_WALLET" | "CASH" | "SAVINGS";
export type AccountStatus = "ACTIVE" | "LOCKED" | "CLOSED" | "EXPIRED" | "REPLACED";

export interface Account {
  id: string;
  institution_id?: string;
  account_name: string;
  account_type: AccountType;
  card_number_masked: string;
  card_number_last4: string;
  initial_balance?: number;
  credit_limit: number;
  billing_day_of_month?: number;
  grace_period_days: number;
  status: AccountStatus;
  replaces_account_id?: string;
  opened_date?: string;
  closed_date?: string;
  color_hex?: string;
  note?: string;
  created_at?: string;
  updated_at?: string;
  institution?: Institution;
}

export interface AccountOverview {
  account_id: string;
  account_name: string;
  bank_name: string;
  card_number_masked: string;
  credit_limit: number;
  latest_statement_balance: number;
  next_payment_due_date: string;
  status: AccountStatus;
  replaces_account_id?: string;
  replaced_by_card_number?: string;
}

export interface AccountLiveBalance {
  account_id: string;
  account_name: string;
  account_type?: AccountType;
  is_asset?: boolean;
  bank_name?: string;
  card_number_masked?: string;
  color_hex?: string;
  initial_balance?: number;
  credit_limit: number;
  latest_statement_date?: string;
  latest_statement_balance: number;
  unbilled_charges: number;
  unbilled_credits: number;
  unbilled_net_amount: number;
  unbilled_transaction_count: number;
  live_current_balance: number;
  live_available_limit: number;
  live_utilization_percentage: number;
  live_risk_level: string;
  next_payment_due_date?: string;
  status: AccountStatus;
}

export interface AccountCreatePayload {
  institution_id?: string;
  account_name: string;
  account_type?: AccountType;
  card_number_masked?: string;
  card_number_last4?: string;
  initial_balance?: number;
  credit_limit?: number;
  billing_day_of_month?: number;
  grace_period_days?: number;
  status?: AccountStatus;
  replaces_account_id?: string;
  color_hex?: string;
  note?: string;
}

export interface AccountUpdatePayload {
  account_name?: string;
  initial_balance?: number;
  credit_limit?: number;
  billing_day_of_month?: number;
  grace_period_days?: number;
  status?: AccountStatus;
  replaces_account_id?: string;
  closed_date?: string;
  color_hex?: string;
  note?: string;
}
